In [2]:
from numpy import array
from scipy.linalg import null_space, det, norm, inv

# this loads some HTML style files
from IPython.core.display import HTML
with open( './style/custom.css', 'r' ) as f: html_style = f.read()
HTML( html_style )

### Domanda 1

Risolvere il seguente sistema

$$
\left\{
\begin{matrix}
  3x+ 6y+ 3z+ 3w & =& 1\\
  2x+ 9y+ 7z+ 3w & =& 2\\
  9x+ 8y+ 5z & =& 3\\
  9x+ 8y+ 7z & =& 3\\
\end{matrix}
\right.
$$

### Risposta 1

L'equazione matriciale associata è

$$
\begin{bmatrix}
  3& 6& 3& 3\\
  2& 9& 7& 3\\
  9& 8& 5& 0\\
  9& 8& 7& 0\\
\end{bmatrix}
\cdot
\begin{bmatrix}
  x\\ y\\ z\\ w\\
\end{bmatrix}
\quad=\quad
\begin{bmatrix}
  1\\ 2\\ 3\\ 3\\
\end{bmatrix}
$$

Se la matrice non è singolare, la soluzione è

$$
\begin{bmatrix}
  x\\ y\\ z\\ w\\
\end{bmatrix}
\quad=\quad
\begin{bmatrix}
  3& 6& 3& 3\\
  2& 9& 7& 3\\
  9& 8& 5& 0\\
  9& 8& 7& 0\\
\end{bmatrix}^{-1}
\cdot
\begin{bmatrix}
  1\\ 2\\ 3\\ 3\\
\end{bmatrix}
$$


In [4]:
A = array ( [ [3,6,3,3], [2,9,7,3], [9,8,5,0], [9,8,7,0] ] )
b = array ( [ [1,2,3,3] ] ).T
inv(A) @ b

array([[ 0.02857143],
       [ 0.34285714],
       [ 0.        ],
       [-0.38095238]])

### Domanda 2

Trovare tutte le soluzioni del sistema

$$
\left\{
\begin{matrix}
  3x+ 6y+ 3z+ 3w & =& 0\\
  2x+ 9y+ 7z+ 3w & =& 0\\
  9x+ 8y+ 5z & =& 0\\
  x- 3y - 4z & =& 0\\
\end{matrix}
\right.
$$

### Risposta 2

L'equazione matriciale associata è

$$
\begin{bmatrix}
  3& 6& 3& 3\\
  2& 9& 7& 3\\
  9& 8& 5& 0\\
  1& -3& -4& 0\\
\end{bmatrix}
\cdot
\begin{bmatrix}
  x\\ y\\ z\\ w\\
\end{bmatrix}
\quad=\quad
\begin{bmatrix}
  0\\ 0\\ 0\\ 0\\
\end{bmatrix}
$$


In [9]:
A = array ( [ [3,6,3,3], [2,9,7,3], [9,8,5,0], [1,-3,-4,0] ]  )
null_space(A)

array([[-0.26565743],
       [ 0.64070322],
       [-0.54694177],
       [-0.46880723]])

Le soluzioni sono tutti i vettori $c\cdot[-0.27 ,0.64, -0.55, -0.47]$ per un arbitrario scalare $c$.

### Domanda 3

Trovare tutte le soluzioni del sistema

$$
\left\{
\begin{matrix}
  3x+ 6y+ 3z+ 3w & =& 0\\
  2x+ 9y+ 7z+ 3w & =& 0\\
  5x+ 15y+ 10z+ 6w & =& 0\\
  x- 3y - 4z & =& 0\\
\end{matrix}
\right.
$$

### Risposta 3

L'equazione matriciale associata è

$$
\begin{bmatrix}
  3& 6& 3& 3\\
  2& 9& 7& 3\\
  5& 15& 10& 6\\
  1& -3& -4& 0\\
\end{bmatrix}
\cdot
\begin{bmatrix}
  x\\ y\\ z\\ w\\
\end{bmatrix}
\quad=\quad
\begin{bmatrix}
  0\\ 0\\ 0\\ 0\\
\end{bmatrix}
$$


In [5]:
A = array ( [ [3,6,3,3], [2,9,7,3], [5,15,10,6], [1,-3,-4,0] ]  )
null_space(A)

array([[-0.57874361,  0.40013165],
       [-0.06705467, -0.64136203],
       [-0.0943949 ,  0.58105444],
       [ 0.80724786,  0.30153798]])

Le soluzioni sono tutti i vettori 

$c\cdot[-0.58 ,-0.07, -0.09, 0.81]\ +\ d\cdot[0.4 ,-0.64, 0.58, 0.3]$ 

per un arbitraria coppia di scalari $c,d$.